src/recursive_shared_mutex.cpp

```diff
@@ -0,0 +1,338 @@
+                                   "exclusive ownership");
+        }
+        _shared_while_exclusive_counter--;
+        return true;
+    }
+    return false;
+}
+
+bool recursive_shared_mutex::already_has_lock_shared(const std::thread::id &locking_thread_id)
+{
+    return (_read_owner_ids.find(locking_thread_id) != _read_owner_ids.end());
+}
+
+void recursive_shared_mutex::lock_shared_internal(const std::thread::id &locking_thread_id, const uint64_t &count)
+{
+    auto it = _read_owner_ids.find(locking_thread_id);
+    if (it == _read_owner_ids.end())
+    {
+        _read_owner_ids.emplace(locking_thread_id, count);
+    }
+    else
+    {
+        it->second = it->second + count;
+    }
+}
+
+void recursive_shared_mutex::unlock_shared_internal(const std::thread::id &locking_thread_id, const uint64_t &count)
+{
+    auto it = _read_owner_ids.find(locking_thread_id);
+    if (it == _read_owner_ids.end())
+    {
+        throw std::logic_error("can not unlock_shared more times than we locked for shared ownership");
+    }
+    it->second = it->second - count;
+    if (it->second == 0)
+    {
+        _read_owner_ids.erase(it);
+    }
+}
+
+uint64_t recursive_shared_mutex::get_shared_lock_count(const std::thread::id &locking_thread_id)
+{
+    auto it = _read_owner_ids.find(locking_thread_id);
+    if (it == _read_owner_ids.end())
+    {
+        return 0;
+    }
+    return it->second;
+}
+
+void recursive_shared_mutex::lock_auto_locks(const std::thread::id &locking_thread_id, const uint64_t &count)
+{
+    if (_auto_unlock_id != NON_THREAD_ID)
+    {
+        throw std::logic_error("lock_auto_locks incorrectly called while already occupied by another thread");
+    }
+    _auto_unlock_id = locking_thread_id;
+    _auto_unlock_count = count;
+}
+
+uint64_t recursive_shared_mutex::get_auto_lock_count(const std::thread::id &locking_thread_id)
+{
+    if (_auto_unlock_id == locking_thread_id)
+    {
+        return _auto_unlock_count;
+    }
+#ifdef DEBUG
+    throw std::logic_error("get_auto_lock_count incorrectly called on a thread with no auto locks");
+#else
+    return 0;
+#endif
+}
+
+void recursive_shared_mutex::unlock_auto_locks(const std::thread::id &locking_thread_id)
+{
+    if (_auto_unlock_id == locking_thread_id)
+    {
+        _auto_unlock_count = 0;
+        _auto_unlock_id = NON_THREAD_ID;
+        return;
+    }
+    throw std::logic_error("unlock_auto_locks incorrectly called on a thread with no auto locks");
+}
+
+////////////////////////
+///
+/// Public Functions
+///
+
+void recursive_shared_mutex::lock()
+{
+    const std::thread::id &locking_thread_id = std::this_thread::get_id();
+    std::unique_lock<std::mutex> _lock(_mutex);
+    if (_write_owner_id == locking_thread_id)
+    {
+        _write_counter++;
+    }
+    else
+    {
+        // Wait until we can set the write-entered.
+        _read_gate.wait(_lock, [=] { return end_of_exclusive_ownership(); });
```

Comment:
I think that there are a few reasons to not use the full capture by value [=] in these lambdas, although the code works.

    efficiency: you are making copies of unnecessary variables
    you are copying _lock which might have weird semantics
    it almost looks like a bug because = captures by value and so a surface reading would imply that _write_promotion_counter should not change. It works because _write_promotion_counter is a member variable and you are really capturing "this" by value and so effectively _write_promotion_counter by reference. But this is quite confusing

Instead why not use [this]?

This comment applies to most or maybe all of your lambdas...

Reply:
[=] is what is used by std::recursive_mutex but you are right about the unnecessary copies and it turns out [this] is slightly more performant.  i have adjusted them to use [this]